test/cpp/tensorexpr/test_cpp_codegen.cpp

```diff
@@ -1,57 +1,296 @@
+  STR_CHECK(cast, "static_cast<float>(1)");
+}
+
+TEST(CppPrinter, BitCast) {
+  KernelScope kernel_scope;
+  auto cast = new BitCast(kInt, new FloatImm(20));
+  STR_CHECK(cast, "std::bitcast<float, int>(20.f)");
+}
+
+TEST(CppPrinter, Let) {
+  KernelScope kernel_scope;
+  auto var = new Var("x", kFloat);
+  auto val = new FloatImm(2);
+  auto let = new Let(var, val);
+  STR_CHECK(let, "float x = 2.f;\n");
+}
 
-  std::stringstream ss;
-  CppPrinter printer(&ss);
-  printer.visit(block);
-  // size(long) = 8;
-  // dim0 * dim1 * dim2 * size(long) = 24000.
-  const std::string expected = R"(
-    # CHECK: {
-    # CHECK:   int64_t* y = static_cast<int64_t*>(malloc(24000));
-    # CHECK:   free(y);
+TEST(CppPrinter, For) {
+  KernelScope kernel_scope;
+  constexpr int N = 1024;
+  Placeholder a(BufHandle("A", {N}, kInt));
+  Placeholder b(BufHandle("B", {N}, kInt));
+  Placeholder c(BufHandle("C", {N}, kInt));
+  VarHandle i("i", kInt);
+  auto f = For::make(i, 0, N, c.store({i}, Add::make(a.load(i), b.load(i))));
+  const std::string pattern = R"(
+   # CHECK: for (int i = 0; i < 1024; i++) {
+   # CHECK:   C[i] = (A[i]) + (B[i]);
+   # CHECK: }
+  )";
+  FILE_CHECK(f, pattern);
+}
+
+TEST(CppPrinter, Cond) {
+  KernelScope kernel_scope;
+  Placeholder x(BufHandle("X", {1}, kInt));
+  auto cmp = CompareSelect::make(x.load(0), 10, CompareSelectOperation::kLT);
+  auto cond =
+      Cond::make(cmp, x.store({0}, x.load(0) + 1), x.store({0}, x.load(0) - 1));
+  const std::string pattern = R"(
+    # CHECK: if (((X[0] < 10) ? 1 : 0)) {
+    # CHECK:   X[0] = (X[0]) + 1;
+    # CHECK: } else {
+    # CHECK:   X[0] = (X[0]) - 1;
     # CHECK: }
   )";
-  torch::jit::testing::FileCheck().run(expected, ss.str());
+  FILE_CHECK(cond, pattern);
+}
+
+TEST(CppPrinter, Intrinsics) {
+  KernelScope kernel_scope;
+  const std::unordered_set<IntrinsicsOp, std::hash<int>> unsupported_ops{
+      kRand, kSigmoid};
+  for (int i = 0; i < kMaxIntrinsicsOp; i++) {
+    IntrinsicsOp op = static_cast<IntrinsicsOp>(i);
+    if (unsupported_ops.count(op)) {
+      continue;
+    }
+
+    if (Intrinsics::OpArgCount(op) == 1) {
+      auto v = new Intrinsics(op, new FloatImm(2.0f));
+      STR_CHECK(v, "std::" + v->func_name() + "(2.f)");
+    } else {
+      auto v = new Intrinsics(op, new FloatImm(1.0f), new FloatImm(2.0f));
+      STR_CHECK(v, "std::" + v->func_name() + "(1.f, 2.f)");
+    }
+  }
+}
+
+TEST(CppPrinter, ExternalCall) {
+  KernelScope kernel_scope;
+  Buf* output = new Buf("out", {new IntImm(2), new IntImm(2)}, kFloat);
+  Buf* buf_arg1 = new Buf("a", {new IntImm(2), new IntImm(2)}, kFloat);
+  Buf* buf_arg2 = new Buf("b", {new IntImm(2), new IntImm(2)}, kFloat);
+  Expr* scalar_arg = new Add(new IntImm(1), new IntImm(2));
+  auto call = new ExternalCall(
+      output, "nnc_aten_matmul", {buf_arg1, buf_arg2}, {scalar_arg});
+  const std::string pattern = R"(
+   # CHECK: {
+   # CHECK:   std::vector<void*> buf_ptrs{out, a, b};
```

Comment:
Minor: we probably could use raw C arrays here and avoid dependency on `<vector>`